### DISCUSSION TITLE
How to handle background notifications in flutter web?

### DISCUSSION THREAD

**Discussion by ChetanKhanna767** [firebase/flutterfire#10294] [Q&A] [answered]
Service worker file is listening to background notificaions, they are printed on console. Tapping on them also works, event is getting printed on "notificationclick" event

```
firebase.initializeApp({
      apiKey: jsonResult['apiKey'],
      appId: jsonResult['appId'],
      messagingSenderId: jsonResult['messagingSenderId'],
      projectId: jsonResult['projectId'],
      authDomain: jsonResult['authDomain'],
      storageBucket: jsonResult['storageBucket'],
      measurementId: jsonResult['measurementId'],
    });

  const messaging = firebase.messaging();
   
  messaging.onBackgroundMessage(function(payload) {
    console.log('Received background message ', payload);
    var current = new Date();
     
  });

function handleClick (event) {

  console.log("onNotificationClickevent", event);
   
}
self.addEventListener('notificationclick', handleClick);
```
Problem is I want to listen to this notification in flutter, firebase messaging listeners don't work in this case. I want to open a new screen with some new data

![Screenshot 2023-01-19 041239](https://user-images.githubusercontent.com/101747174/213311741-fd3b2152-1cbd-4c1a-84f5-b5a87b49c148.png)

**Comment by toco1001** [comment]
Any updates? 

@ChetanKhanna767 
Did you resolve this?

**Reply by ChetanKhanna767** [reply, marked as answer]
Hi, you can follow these steps
1. In service worker , post channel messages on receiving fcm

```
const channel = new BroadcastChannel('sw-messages');

function handleClick(event) {
  console.log("onNotificationClickeventEvent",  event);

  if('FCM_MSG' in event.notification.data)
  {
    console.log("onNotificationClickevent", {'fcm_data': event.notification.data.FCM_MSG, 'clicked' : 'true'});
    channel.postMessage(
      {'fcm_data': event.notification.data.FCM_MSG, 'clicked' : 'true'}
    );
  }
}

 messaging.onBackgroundMessage(function (payload) {

            console.log('Received background message ', payload);
                  
            channel.postMessage(
              payload
            );
    
          })

```
2. You will receive this in html 
 
```
    const channel = new BroadcastChannel('sw-messages');
    channel.addEventListener("message", function (event) {
    localStorage.setItem('flutter.notificationExist', true)
     window.postMessage(JSON.stringify(event.data))

    })
```
3. Listen to these messages in flutter, for this you can use flutter html object or this [package](https://pub.dev/packages/onmessage)
```
   OnMessage.instance.stream.listen((MessageEvent event) {
      final data = jsonDecode(event.data.toString());
  
    });
```

**Reply by toco1001** [reply]
ChetanKhanna767, Thank you for sharing!!
I'll try them.

**Reply by i84762** [reply]
Worked perfectly, thanks a lot.

**Reply by Refaydeen** [reply]
I need to push a notification as you mentioned a image from flutter how to do that ,is that possible to push a notification like that .

**Reply by dsyrstad** [reply]
For those coming here and trying to get the above solution to work, the `notificationclick` event listener must be added _prior_ to importing the firebasejs libraries:

```js
function handleClick(event) {
  console.log("onNotificationClick Event",  event);

  if ('FCM_MSG' in event.notification.data) {
    console.log("FCM onNotificationClick event", {'fcm_data': event.notification.data.FCM_MSG, 'clicked' : 'true'});
    // channel.postMessage(
    //     {'fcm_data': event.notification.data.FCM_MSG, 'clicked' : 'true'}
    // );
  }
}

self.addEventListener('notificationclick', handleClick);

importScripts("https://www.gstatic.com/firebasejs/9.10.0/firebase-app-compat.js");
importScripts("https://www.gstatic.com/firebasejs/9.10.0/firebase-messaging-compat.js");

firebase.initializeApp({
...
```

If you try to set if _after_ firebase initialization, the `handleClick` function won't be called.

**Reply by MartinDawson** [reply]
@dsyrstad Also just an FYI, it doesn't work on localhost:

> Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/notificationclick_event#browser_compatibility).

**Comment by Trung15010802** [comment]
How can i create a "service worker " in flutter? Can you tell me please? @ChetanKhanna767

**Reply by ChetanKhanna767** [reply]
Create a file named "frebase-messaging-sw.js" inside web folder. There's a lot of documentation regarding on how to create service worker in flutter web. You should check that as well

**Reply by antuanjoseff** [reply]
Thanks ChetanKhanna767
I works like a charm!
I owe you big time

**Comment by AbhayJadoun** [comment]
how to receive click event in flutter code
